### DISCUSSION TITLE
Compatibility of modules from independent builds

### DISCUSSION THREAD

**Discussion by kcoop** [parcel-bundler/parcel#8323] [Q&A] [answered]
I’m looking into adding extensibility to a browser based app. The plug-ins themselves would use Typescript + React + MUI, and the app would expose a core module of its own to the plugins. 

A bit bundler ignorant, wondering how this could work with Parcel. Dynamic loading seems promising for the extension modules, but how compatible would a production build be with extensions built independently? Things like HMR seem problematic too?

**Comment by mischnic** [comment, marked as answer]
We have https://github.com/parcel-bundler/parcel/pull/5239 so that multiple projects can run in the same JS context and not interfere (but this also means that shared code including React and MUI will be downloaded and executed twice, etc.)

One solution would be to just not use any ES module functionality but make the core app set some global function which is then called by plugins to register themselves.

With the above PR, these components are completely separate apps from Parcel's perspective so HMR would work in that sense. But you'd have to figure out a way to notify the app that some extension was updated and then do something to apply the change (depends on what your plugins can do).

**Reply by kcoop** [reply]
Thanks, that's clarifying. It sounds like duplicated code means any API will have to be strongly decoupled, so for example extending core classes from within plugins would be problematic. From a systems pov this is desirable anyway I suppose.

**Reply by mischnic** [reply]
You might be interested in [module federation](https://webpack.js.org/concepts/module-federation/)

**Reply by kcoop** [reply]
Thanks for the pointer, looks like exactly what I’d envisioned. Maybe we’re headed for the lovely world of webpack…
